shared/middleware/varnish_environment.rb

```diff
@@ -1,34 +1,77 @@
 require 'sinatra/base'
 
+require 'cdo/i18n'
+require 'cdo/shared_constants'
+require 'cdo/rack/global_edition'
+require 'helpers/cookies'
+
+require_relative '../../dashboard/lib/metrics/events' # rubocop:disable CustomCops/DashboardRequires
+
 class VarnishEnvironment < Sinatra::Base
+  LOCALE_PARAM_KEY = 'set_locale'.freeze
+
   def self.load_supported_locales
-    Dir.glob(pegasus_dir('cache', 'i18n', '*.json')).map do |i|
-      File.basename(i, '.json').downcase
-    end.sort
+    ::Cdo::I18n.available_languages.map {|cdo_language| cdo_language[:locale_s].downcase}.sort
   end
 
   configure do
     set :locales_supported, load_supported_locales
   end
 
   before do
-    env['cdo.locale'] = varnish_locale || cookie_locale || default_locale
+    env['cdo.locale'] = param_locale || varnish_locale || cookie_locale || default_locale
   end
 
   after do
+    if param_locale
+      set_locale_cookie(param_locale)
+
+      redirect_uri = URI(request.path)
+      redirect_params = request.params.except(LOCALE_PARAM_KEY)
+
+      unless param_ge_region == request.ge_region
+        redirect_params[Rack::GlobalEdition::REGION_KEY] = param_ge_region
+
+        Metrics::Events.log_event(
+          event_name: 'Global Edition Region Selected',
+          request: request,
+          metadata: {
+            country: request.country_code,
+            region: param_ge_region,
+            locale: param_locale,
+          }
+        )
+      end
+
+      redirect_uri.query = URI.encode_www_form(redirect_params).presence
+
+      response.redirect(redirect_uri.to_s)
+    end
   end
 
   helpers do
+    include Middleware::Helpers::Cookies
+
     def varnish_locale
       language_to_locale(request.env['HTTP_X_VARNISH_ACCEPT_LANGUAGE'])
     end
 
     def cookie_locale
-      language_to_locale(request.cookies['language_'])
+      language_to_locale(locale_cookie)
```

Comment:
It feels confusing to have both a helper named `cookie_locale` and one it uses named `locale_cookie` ... maybe we can have a different naming pattern. Perhaps `locale_from_cookie` and `locale_from_varnish`? or `cookie_locale_value` and `varnish_locale_value`... not sure.

I think they are only ever really used in the middleware config, so hopefully they are not a pain to name?

Reply:
https://github.com/code-dot-org/code-dot-org/pull/62061/commits/0cecaf1bd36313442200e8a1a20d0f1c3eb40228